### ISSUE TITLE
how to interpret autoanchor log and improve more BPR

### ISSUE THREAD

**Issue by YoungjaeDev** [ultralytics/yolov5#7118]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello, I train with custom dataset, and it has various data size. 
Therefore, I set it to 4 while using the annotation of anchor from hyp because I wanted to customize it and get higher mAP  performance. But I have a question about the autoanchor log.

```bash
AutoAnchor: 0.48 anchors/target, 0.114 Best Possible Recall (BPR). Anchors are a poor fit to dataset ⚠️, attempting to improve...
AutoAnchor: WARNING: Extremely small objects found: 256 of 129021 labels are < 3 pixels in size
AutoAnchor: Running kmeans for 12 anchors on 129018 points...
AutoAnchor: Evolving anchors with Genetic Algorithm: fitness = 0.7203: 100%|██████████| 1000/1000 [00:06<00:00, 154.38it/s]                                                                    
AutoAnchor: thr=0.25: 0.9962 best possible recall, 5.13 anchors past thr
AutoAnchor: n=12, img_size=960, metric_all=0.266/0.721-mean/best, past_thr=0.468-mean: 21,14, 40,27, 86,24, 57,54, 131,52, 94,143, 238,79, 229,171, 526,118, 305,442, 664,274, 794,574
AutoAnchor: Done ✅ (optional: update model *.yaml to use these anchors in the future)
```

1. what is Evolving anchors with Genetic Algorithm: fitness (0.7203)? Is 0.7203 an appropriate value?
2. thr=0.25, Does thr mean the iou threshold of gt and anchor?
3. What is past_ptr? So it's confusing what the value of 5.13 anchor is also available. In the end, it is questionable whether anchor should be reset to 5.
4. In the end, it seems to be going to 4 anchor now, how do you improve more than the present when you look at the log?

Thank you

**Comment by glenn-jocher** [comment]
@youngjae-avikus 👋 Hello! Thanks for asking about model anchors. YOLOv5 🚀 uses a new [Ultralytics](https://ultralytics.com/) algorithm called **AutoAnchor** for anchor verification and generation before training starts.

Autoanchor will analyse your anchors against your dataset and training settings (like `--img-size`), and will adjust your anchors as necessary if it determines the original anchors are a poor fit, or if an **anchor count** was specified in your model.yaml rather than **anchor values**, i.e. 
```yaml
# Specify anchor count (per layer)
anchors: 3

# --OR-- Specify anchor values manually
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```

When generating new anchors, autoanchor first applies a kmeans function against your dataset labels (scaled to your training `--img-size`), and uses kmeans centroids as initial conditions for a [Genetic Evolution](https://en.wikipedia.org/wiki/Genetic_algorithm) (GE) algorithm. The GE algorithm will evolve all anchors for 1000 generations under default settings, using CIoU loss (same regression loss used during training) combined with [Best Possible Recall](https://en.wikipedia.org/wiki/Precision_and_recall) (BPR) as its fitness function.

Notebook example: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img width="1200" alt="Screenshot 2022-03-24 at 10 58 28" src="https://user-images.githubusercontent.com/26833433/159891298-8b15f80b-f5b0-4fa9-8824-7380bae32a76.png">

**No action is required on your part** to use autoanchor. If you would like to force manual anchors for any reason, you can skip autoanchor with the `--noautoanchor` flag:
```shell
python train.py --noautoanchor
```

For more details on AutoAnchor see:
https://github.com/ultralytics/yolov5/blob/master/utils/autoanchor.py

Good luck 🍀 and let us know if you have any other questions!

**Comment by YoungjaeDev** [comment]
@glenn-jocher  Thank you. Additionally, I would like to know my log **Autoanchor 5.13 anchors past thr**. Does it mean that BPR is 0.99 when set to Anchor 5?

**Comment by glenn-jocher** [comment]
@youngjae-avikus anchors past threshold counts the number of anchors that match a label on average for your dataset.

**Comment by YoungjaeDev** [comment]
@glenn-jocher When I read the code related check_anchor, I don't understand how it comes out because I don't understand evolve side, but in conclusion, did `past-threshold` say 5 as an anchor that **overfits** my dataset? Or does 5 mean 5anchors/target? Thank you

**Comment by glenn-jocher** [comment]
5 anchors per target

**Comment by YoungjaeDev** [comment]
@glenn-jocher 

The image data changes when fine-tuning, but the number of anchor boxes doesn't have to be the same (eg. coco) as before, so it's better to turn on the anchor option, is it right?

**Comment by glenn-jocher** [comment]
@youngjae-avikus yes, that's correct. It's often helpful to tune the anchor options when fine-tuning on a custom dataset with different characteristics than the original dataset. This can help optimize performance for your specific use case.
